File: apps/api/apps/librarian/routes/inventory.js
const { fetchItems, fetchItem, createItems, updateItem, deleteItem, convertItem } = require('../../../services/inventory');

const express = require('express');
const router = express.Router();

router.get('/', async (req, res) => {
    try {
        res.send(await fetchItems());
    } catch (error) {
        console.error(error);
        res.status(error.status || 500).send({ errors: [error] });
    }
});

router.get('/:id', async (req, res) => {
    try {
        res.send(await fetchItem(req.params.id));
    } catch (error) {
        console.error(error);
        res.status(error.status || 500).send({ errors: [error] });
    }
});

router.put('/', async (req, res) => {
    const { thingId, quantity, brand, condition, notes, estimatedValue, hidden, image, manuals } = req.body;

    try {
        res.send(await createItems(thingId, { quantity, brand, condition, notes, estimatedValue, hidden, image, manuals }));
    } catch (error) {
        console.error(error);
        res.status(error.status || 500).send({ errors: [error] });
    }
});

router.patch('/:id', async (req, res) => {
    const { brand, notes, estimatedValue, hidden, condition, image, manuals } = req.body;

    try {
        await updateItem(req.params.id, { brand, notes, estimatedValue, hidden, condition, image, manuals });
        res.status(204).send();
    } catch (error) {
        console.error(error);
        res.status(error.status || 500).send({ errors: [error] });
    }
});

router.post('/:id/convert', async (req, res) => {
    const { id } = req.params;
    const { thingId } = req.body;

    try {
        await convertItem(id, thingId);
        res.status(204).send();
    } catch (error) {
        console.error(error);
        res.status(error.status || 500).send({ errors: [error] });
    }
});

router.delete('/:id', async (req, res) => {
    const { id } = req.params;

    try {
        await deleteItem(id);
        res.status(204).send();
    } catch (error) {
        console.error(error);
        res.status(error.status || 500).send({ errors: [error] });
    }
});

module.exports = router;

File: apps/librarian/lib/core/api/inventory.dart
part of 'api.dart';

Future<Response> fetchInventoryItems() async {
  return await DioClient.instance.get('/inventory');
}

Future<Response> fetchInventoryItem({required int number}) async {
  return await DioClient.instance.get('/inventory/$number');
}

Future<Response> createInventoryItems(
  String thingId, {
  required int quantity,
  required String? brand,
  String? condition,
  required String? notes,
  required double? estimatedValue,
  bool? hidden,
  ImageDTO? image,
  List<ImageDTO>? manuals,
}) async {
  return await DioClient.instance.put('/inventory', data: {
    'thingId': thingId,
    'quantity': quantity,
    'brand': brand,
    'condition': condition,
    'notes': notes,
    'estimatedValue': estimatedValue,
    'hidden': hidden,
    'image': {
      'url': image?.url,
    },
    'manuals': manuals?.map((m) => {'url': m.url, 'filename': m.name}).toList(),
  });
}

Future<Response> updateInventoryItem(
  String id, {
  String? brand,
  String? condition,
  String? notes,
  double? estimatedValue,
  bool? hidden,
  ImageDTO? image,
  List<ImageDTO>? manuals,
}) async {
  return await DioClient.instance.patch('/inventory/$id', data: {
    'brand': brand,
    'condition': condition,
    'notes': notes,
    'estimatedValue': estimatedValue,
    'hidden': hidden,
    'image': image != null ? {'url': image.url} : null,
    'manuals': manuals?.map((m) => {'url': m.url, 'filename': m.name}).toList(),
  });
}

Future<Response> convertInventoryItem(
  String id,
  String thingId,
) async {
  return await DioClient.instance.post('/inventory/$id/convert', data: {
    'thingId': thingId,
  });
}

Future<Response> deleteInventoryItem(String id) async {
  return await DioClient.instance.delete('/inventory/$id');
}
